### PR TITLE
add Decoder / Encoder interfaces

### PR DESCRIPTION
**Pull request by gcanti**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ declare global {
 
 export interface ContextEntry {
   readonly key: string
-  readonly type: Any | NeverType
+  readonly type: Decoder<any, any>
```
Comment (sledorze) [inline src/index.ts]: @gcanti do we loose anything by dropping 'never' ?

Comment (gcanti) [reply src/index.ts]: @sledorze `| NeverType` was added for variance issues, IMO this is an actual improvement due to the Decoder/Encoder splitting

Comment (sledorze) [reply src/index.ts]: @gcanti  ok good to know 👍

Comment (gcanti) [reply src/index.ts]: For reference, these are the lines that currently cause variance issues if I remove `| NeverType`

https://github.com/gcanti/io-ts/blob/37c74a5e2038de063a950f9ba8d18b1f132ef450/src/index.ts#L756
https://github.com/gcanti/io-ts/blob/37c74a5e2038de063a950f9ba8d18b1f132ef450/src/index.ts#L875

**Comment by gcanti** [comment]
@sledorze rebased on top of master, should I release a next version after merging this?

**Comment by sledorze** [comment]
@gcanti yes please, I may integrate it over le week-end.

**Comment by gcanti** [comment]
@sledorze published as `io-ts@0.9.2-dev.20171215`. Let me know how it goes

**Comment by sledorze** [comment]
@gcanti it works perfectly! The code involved looks nicer too!
(Note: I had to manually change io-ts-types ref to io-ts to prevent type duplication)

**Pull request merged**